Implement snapshot query trait

christianwwwwwwww opened the pull request:
This PR adds a new `SnapshotQuery` struct to the SDK wasm and hooks it into the Program Manager's execute.rs as a replacement for `QueryNative`. 

Previously the query execution was consistently failing due to `stateRoot` mismatches happening as the query process was often slower than the ~2 second block time.

christianwwwwwwww committed:
implement snapshot query trait

christianwwwwwwww committed:
create snapshot query helpers

christianwwwwwwww committed:
reorganize

christianwwwwwwww committed:
collect commitments and fetch statePaths concurrently

christianwwwwwwww committed:
view_key naming

iamalwaysuncomfortable commented on wasm/src/programs/snapshot_query.rs:
```diff
@@ -0,0 +1,202 @@
+/* --------------------------- internal helpers --------------------------- */
+
+/// Heuristic: detect plaintext records from JS inputs and compute commitments.
+fn collect_commitments_from_inputs(
```
These should be part of the object itself.

iamalwaysuncomfortable commented on wasm/src/programs/snapshot_query.rs:
```diff
@@ -0,0 +1,202 @@
+        node_url: &str,
+        program_id: &ProgramID<CurrentNetwork>,
+        record_name: &Identifier<CurrentNetwork>,
+        view_key: Field<CurrentNetwork>,
```
```suggestion
        view_key: &ViewKeyNative<CurrentNetwork>,
```

iamalwaysuncomfortable commented on wasm/src/programs/snapshot_query.rs:
```diff
@@ -0,0 +1,202 @@
+fn collect_commitments_from_inputs(
+    program_id: &ProgramIDNative,
+    record_name: &IdentifierNative,
+    view_key: Field<CurrentNetwork>,
```
```suggestion
    view_key: &ViewKeyNative<CurrentNetwork>,
```

iamalwaysuncomfortable commented on wasm/src/programs/snapshot_query.rs:
```diff
@@ -0,0 +1,202 @@
+            }
+            // Try parse as a plaintext record. Skip if not a record.
+            if let Ok(rec) = RecordPlaintextNative::from_str(&s) {
+                let cm = rec.to_commitment(program_id, record_name, &view_key);
```
```suggestion
                let record_view_key = (&**view_key * &self.nonce).to_x_coordinate();
                let commitment = record.to_commitment(program_id, record_name, &record_view_key);
```

Note here that the record view key is a different thing from the view key of the account. Each record has its own view key, the above is how you compute it.

iamalwaysuncomfortable commented on wasm/src/programs/snapshot_query.rs:
```diff
@@ -0,0 +1,202 @@
+    // Example (pseudo):
+    // client.state_path_at_root(_node_url, _commitment, _state_root).await
+    Err(anyhow!("fetch_state_path_at_root() not implemented"))
+}
```
Any helpers should be inside the struct.

iamalwaysuncomfortable commented on wasm/src/programs/snapshot_query.rs:
```diff
@@ -0,0 +1,202 @@
+        for (cm, res) in commitments.iter().zip(results.into_iter()) {
+            let sp_str = res?;
+            // Optional safety: parse and ensure the path's root matches the pinned root.
+            let sp = StatePath::<CurrentNetwork>::from_str(&sp_str).map_err(|e| anyhow!(e.to_string()))?;
```
Don't abbreviate the variable names, cm should be commitment, sp should be state_path, etc.

iamalwaysuncomfortable reviewed (commented):
A couple guiding comments for this PR.

christianwwwwwwww committed:
address comments

christianwwwwwwww committed:
point snarkvm dependencies to 4.2.0 rev

iamalwaysuncomfortable committed:
Add debug mode to network client

iamalwaysuncomfortable committed:
Update SnapshotQuery to use new batch statepaths fetching method + upgrade to SnarkVM v4.2.0

With SnarkOS version 4.1.0 a new batch statepaths endpoint was added, this PR adds the SnapShot query to fetch directly from this endpoint due to the SnarkVM query's insufficient design.

Further the upgrade to SnarkVM v4.2.0 requires addition of consensus versions as a method argument to verification routines and deployment cost calculations.
To account for this, this PR adds automatic detection of ConsensusVersion and InclusionVersion based on block height detected from the Query.

iamalwaysuncomfortable committed:
Fix function naming error

iamalwaysuncomfortable committed:
Move retries to a constant variable

iamalwaysuncomfortable committed:
Update SnarkVM to working rev 3e94245569ab77724ebde670d7d2685f040e42b8, add tests to REST methods + snapshot query, and fix snapshot query

iamalwaysuncomfortable committed:
Re-add module line in package.json

iamalwaysuncomfortable committed:
Bump release version

iamalwaysuncomfortable committed:
Update active rust version in CI, fix js tests, update SnarkVM to rev e035bec

iamalwaysuncomfortable committed:
Update SDK version to next release version

iamalwaysuncomfortable committed:
Remove multi-threading for CI on template-node test

iamalwaysuncomfortable committed:
Merge pull request #1084 from ProvableHQ/feat/query-trait-extension

[Feature] Snapshot query trait extension for SnarkOS v4.2.1

iamalwaysuncomfortable committed:
Merge branch 'mainnet' into feat/query-trait

iamalwaysuncomfortable committed:
Enable setting verbose transaction errors on the AleoNetworkClient

iamalwaysuncomfortable committed:
Add runner instance size xlarge for template-node test

iamalwaysuncomfortable committed:
Add functionality for decryption of sender ciphertexts

iamalwaysuncomfortable committed:
Update SnarkVM to 4.2.0

iamalwaysuncomfortable committed:
Merge pull request #1089 from ProvableHQ/feat/sender-ciphertext

[Feature] Decrypt Sender Ciphertext

iamalwaysuncomfortable committed:
Update auto-generated api docs
